hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-web-proxy/src/main/java/org/apache/hadoop/yarn/server/webproxy/WebAppProxyServlet.java

```diff
@@ -134,16 +132,19 @@ public WebAppProxyServlet() {
     this.trackingUriPlugins =
         conf.getInstances(YarnConfiguration.YARN_TRACKING_URL_GENERATOR,
             TrackingUriPlugin.class);
-    this.rmAppPageUrlBase =
-        StringHelper.pjoin(WebAppUtils.getResolvedRMWebAppURLWithScheme(conf),
-          "cluster", "app");
     this.failurePageUrlBase =
         StringHelper.pjoin(WebAppUtils.getResolvedRMWebAppURLWithScheme(conf),
           "cluster", "failure");
-    this.ahsAppPageUrlBase =
-        StringHelper.pjoin(WebAppUtils.getHttpSchemePrefix(conf)
-          + WebAppUtils.getAHSWebAppURLWithoutScheme(conf),
-          "applicationhistory", "app");
+  }
+
+  private String getRmAppPageUrlBase(ApplicationId id) throws YarnException, IOException {
+    return ((AppReportFetcher) getServletContext().getAttribute(WebAppProxy.FETCHER_ATTRIBUTE))
```

Comment:
This line of code needs to be refactored, too many layers and not easy to read.

This way of writing might be better.
```
AppReportFetcher appReportFetcher  = (AppReportFetcher) getServletContext();
Attribute attr1 = appReportFetcher.getAttribute(WebAppProxy.FETCHER_ATTRIBUTE);
return attr1.getAhsAppPageUrlBase();
```

Reply:
This code should show like this. If refactor is only for clean code. I think it is not very clean.

```
    ServletContext context = getServletContext();
    AppReportFetcher report =
        ((AppReportFetcher) context.getAttribute(WebAppProxy.FETCHER_ATTRIBUTE));
    return report.getRmAppPageUrlBase(id);
```